src/pytorch/pme/pme.py

```diff
@@ -0,0 +1,194 @@
+        exclusions: torch.Tensor
+            a tensor of shape `(atoms, max_exclusions)` containing excluded interactions, where `max_exclusions` is the
+            maximum number of exclusions for any atom.  Row `i` lists the indices of all atoms with which atom `i` should
+            not interact.  If an atom has less than `max_exclusions` excluded interactions, set the remaining elements
+            in the row to -1.  The exclusions must be symmetric: if `j` appears in row `i`, then `i` must also appear in
+            row `j`.  If you pass a tensor that does not satisfy that requirement, the results are undefined.
+        """
+        if gridx < 1 or gridy < 1 or gridz < 1:
+            raise ValueError('The grid dimensions must be positive')
+        if order < 1:
+            raise ValueError('order must be positive')
+        if alpha <= 0:
+            raise ValueError('alpha must be positive')
+        if coulomb <= 0:
+            raise ValueError('coulomb must be positive')
+        if exclusions.dim() != 2:
+            raise ValueError('exclusions must be 2D')
+        self.gridx = gridx
+        self.gridy = gridy
+        self.gridz = gridz
+        self.order = order
+        self.alpha = alpha
+        self.coulomb = coulomb
+        self.exclusions, _ = torch.sort(exclusions.to(torch.int32), descending=True)
+
+        # Initialize the bspline moduli.
+
+        max_size = max(gridx, gridy, gridz)
+        data = torch.zeros(order, dtype=torch.float32)
+        ddata = torch.zeros(order, dtype=torch.float32)
+        bsplines_data = torch.zeros(max_size, dtype=torch.float32)
+        data[0] = 1
+        for i in range(3, order):
+            data[i-1] = 0
+            for j in range(1, i-1):
+                data[i-j-1] = (j*data[i-j-2]+(i-j)*data[i-j-1])/(i-1)
+            data[0] /= i-1
+
+        # Differentiate.
+
+        ddata[0] = -data[0]
+        ddata[1:order] = data[0:order-1]-data[1:order]
+        for i in range(1, order-1):
+            data[order-i-1] = (i*data[order-i-2]+(order-i)*data[order-i-1])/(order-1)
+        data[0] /= order-1
+        bsplines_data[1:order+1] = data
+
+        # Evaluate the actual bspline moduli for X/Y/Z.
+
+        self.moduli = []
+        for ndata in (gridx, gridy, gridz):
+            m = torch.zeros(ndata, dtype=torch.float32)
+            for i in range(ndata):
+                arg = (2*torch.pi*i/ndata)*torch.arange(ndata)
+                sc = torch.sum(bsplines_data[:ndata]*torch.cos(arg))
+                ss = torch.sum(bsplines_data[:ndata]*torch.sin(arg))
+                m[i] = sc*sc + ss*ss
+            for i in range(ndata):
+                if m[i] < 1e-7:
+                    m[i] = (m[(i-1+ndata)%ndata]+m[(i+1)%ndata])*0.5
+            self.moduli.append(m)
+
+    def compute_direct(self, positions: torch.Tensor, charges: torch.Tensor, cutoff: float, box_vectors: torch.Tensor, max_num_pairs: int = -1):
```

Comment:
The grid size is specified in the constructor, I guess for performance reasons. Does it really make sense then to leave the box size change at every forward call? grid size and box size are kind of tied to each other, right?